File: FactoryMethod/FactoryMethod01.js
// Classe Base de Veiculo
class Veiculo{
    constructor(modelo){
        this.modelo = modelo;
    }

    mostrarDetalhes(){
        console.log(`Modelo: ${this.modelo}`);
    };
};

// Sub de Veiculo
class Carro extends Veiculo {
    constructor(modelo){
        super(modelo);
    };
};

class Moto extends Veiculo {
    constructor(modelo){
        super(modelo);
    };
};

class Barco extends Veiculo {
    constructor(modelo){
        super(modelo);
    };
};


// Fabrica Abstrata de Veiculo
class FabricaDeVeiculos{
    criarVeiculo(modelo){
        throw new Error('O metodo criarVeiculo deve ser implementado');
    };
};

// Fabrica Concreta de Carros 
class FabricaDeCarros extends FabricaDeVeiculos {
    criarVeiculo(modelo){
        return new Carro(modelo);
    };
};

// Fabrica Concreta de Motos 
class FabricaDeMotos extends FabricaDeVeiculos {
    criarVeiculo(modelo){
        return new Moto(modelo);
    };
};

class FabricaDeBarcos extends FabricaDeVeiculos {
    criarVeiculo(modelo){
        return new Barco(modelo);
    };
};

// Uso da Fabrica - Cliente
const fabricaDeCarros = new FabricaDeCarros();
const fabricaDeMotos = new FabricaDeMotos();
const fabricaDeBarcos = new FabricaDeBarcos();

const veiculo1 = fabricaDeCarros.criarVeiculo('Sedan');
const veiculo2 = fabricaDeCarros.criarVeiculo('Hatch');
const veiculo3 = fabricaDeMotos.criarVeiculo('Esportiva');
const veiculo4 = fabricaDeMotos.criarVeiculo('Custon');
const veiculo5 = fabricaDeBarcos.criarVeiculo('Navio MSC');
const veiculo6 = fabricaDeBarcos.criarVeiculo('Lancha 10 pés');

veiculo1.mostrarDetalhes();
veiculo2.mostrarDetalhes();
veiculo3.mostrarDetalhes();
veiculo4.mostrarDetalhes();
veiculo5.mostrarDetalhes();
veiculo6.mostrarDetalhes();

File: FactoryMethod/FactoryMethod01.java
// Classe base de veículo
class Veiculo {
    protected String modelo;

    public Veiculo(String modelo) {
        this.modelo = modelo;
    }

    public void mostrarDetalhes() {
        System.out.println("Veículo modelo: " + modelo);
    }
}
    
// Subclasse Carro
class Carro extends Veiculo {
    public Carro(String modelo) {
        super(modelo);
    }

    public void mostrarDetalhes() {
        System.out.println("Carro modelo: " + modelo);
    }
}

// Subclasse Moto
class Moto extends Veiculo {
    public Moto(String modelo) {
        super(modelo);
    }

    public void mostrarDetalhes() {
        System.out.println("Moto modelo: " + modelo);
    }
}

// Subclasse Barco
class Barco extends Veiculo {
    public Barco(String modelo) {
        super(modelo);
    }

    public void mostrarDetalhes() {
        System.out.println("Barco modelo: " + modelo);
    }
}

// Fábrica abstrata
abstract class FabricaDeVeiculos {
    public abstract Veiculo criarVeiculo(String modelo);
}

// Fábrica concreta de Carros
class FabricaDeCarros extends FabricaDeVeiculos {
    public Veiculo criarVeiculo(String modelo) {
        return new Carro(modelo);
    }
}

// Fábrica concreta de Motos
class FabricaDeMotos extends FabricaDeVeiculos {
    public Veiculo criarVeiculo(String modelo) {
        return new Moto(modelo);
    }
}

// Fábrica concreta de Barcos
class FabricaDeBarcos extends FabricaDeVeiculos {
    public Veiculo criarVeiculo(String modelo) {
        return new Barco(modelo);
    }
}

// Classe principal (Cliente)
public class FactoryMethod01 {
    public static void main(String[] args) {

        FabricaDeVeiculos fabricaDeCarros = new FabricaDeCarros();
        FabricaDeVeiculos fabricaDeMotos = new FabricaDeMotos();
        FabricaDeVeiculos fabricaDeBarcos = new FabricaDeBarcos();

        Veiculo veiculo1 = fabricaDeCarros.criarVeiculo("Sedan");
        Veiculo veiculo2 = fabricaDeMotos.criarVeiculo("Esportiva");
        Veiculo veiculo3 = fabricaDeBarcos.criarVeiculo("Navio");

        veiculo1.mostrarDetalhes();
        veiculo2.mostrarDetalhes();
        veiculo3.mostrarDetalhes();
    }
}